### PR TITLE
Support `foreach list as index, value`

### PR DESCRIPTION
**Pull request by anukul**
Closes https://github.com/storyscript/runtime/issues/585

### CODE REVIEW

File: storyruntime/processing/Lexicon.py [hunk 1]
```diff
@@ -435,8 +435,22 @@ def throw(logger, story, line):
...
+            output = line['output']
+            if len(output) == 1:
+                story.set_variable(assign={'paths': output},
+                                   output=element)
+            elif len(output) == 2:
+                story.set_variable(assign={'paths': [output[0]]},
+                                   output=index)
+                story.set_variable(assign={'paths': [output[1]]},
+                                   output=element)
+            else:
```
Comment (anukul) [inline storyruntime/processing/Lexicon.py]: This branch is a miss because the compiler returns an error and we never parse such a syntax tree.

Comment (wilzbach) [reply storyruntime/processing/Lexicon.py]: Internal assertions should be an `assert`. 
tl;dr:

```py
            else:
                # output length is enforced by the compiler
                assert len(output) == 2
                story.set_variable(assign={'paths': [output[0]]},
                                   output=index)
                story.set_variable(assign={'paths': [output[1]]},
                                   output=element)
```

File: tests/integration/processing/Lexicon.py [hunk 2]
```diff
@@ -471,6 +471,22 @@
...
+                          '  value_sum += value\n',
+        cases=[
+            Case(
+                assertion=ContextAssertion(key='index_sum', expected=3)
+            ),
+            Case(
+                assertion=ContextAssertion(key='value_sum', expected=27)
+            )
+        ]
+    ),
```
Comment (wilzbach) [inline tests/integration/processing/Lexicon.py]: Add a test case for maps

File: tests/integration/processing/Lexicon.py [hunk 3]
```diff
@@ -471,6 +471,38 @@
...
+                          '  value_sum += value\n',
+        cases=[
+            Case(
+                assertion=ContextAssertion(key='key_string', expected='abc')
+            ),
+            Case(
+                assertion=ContextAssertion(key='value_sum', expected=27)
+            )
+        ]
+    ),
```
Comment (wilzbach) [inline tests/integration/processing/Lexicon.py]: There's no test case for dict iteration with a single key.

File: storyruntime/processing/Lexicon.py [hunk 4]
```diff
@@ -433,10 +433,20 @@ def throw(logger, story, line):
...
-        for item in _list:
-            story.set_variable(assign={'paths': line['output']}, output=item)
+        data = story.resolve(line['args'][0], encode=False)
+        assert type(data) in [list, dict], f'Cannot iterate over {type(data)}'
+        iterable = enumerate(data) if isinstance(data, list) else data.items()
+
+        for a, b in iterable:
+            output = line['output']
+            assert 1 <= len(output) <= 2, \
+                f'for loop output must be 1 or 2 values, found {len(output)}'
```
Comment (wilzbach) [inline storyruntime/processing/Lexicon.py]:
```suggestion
                f'foreach loop output must be 1 or 2 values, found {len(output)}'
```

File: storyruntime/processing/Lexicon.py [hunk 5]
```diff
@@ -433,10 +433,20 @@ def throw(logger, story, line):
         """
         Evaluates a for loop.
```
Comment (wilzbach) [inline storyruntime/processing/Lexicon.py]:
```suggestion
        Evaluates a foreach loop.
```

File: storyruntime/processing/Lexicon.py [hunk 6]
```diff
@@ -433,10 +433,20 @@ def throw(logger, story, line):
...
+        assert type(data) in [list, dict], f'Cannot iterate over {type(data)}'
+        iterable = enumerate(data) if isinstance(data, list) else data.items()
+
+        for a, b in iterable:
+            output = line['output']
+            assert 1 <= len(output) <= 2, \
+                f'for loop output must be 1 or 2 values, found {len(output)}'
+            if len(output) == 1:
+                story.set_variable(assign={'paths': output},
+                                   output=b if isinstance(data, list) else a)
```
Comment (wilzbach) [inline storyruntime/processing/Lexicon.py]: Can't we split the code into two parts: one for dict and one for array iteration before doing such evil things? They will bite you later on...

File: tests/integration/processing/Lexicon.py (hunk 3 above)
Comment (anukul) [reply tests/integration/processing/Lexicon.py]: [There is](https://github.com/storyscript/runtime/blob/a61eec8b58a7a3b6072aa3787f6e85268440bded/tests/integration/processing/Lexicon.py#L465), but it's all mixed up in `test_mutation`. I'll fix https://github.com/storyscript/runtime/issues/575 soon.

File: storyruntime/processing/Lexicon.py (hunk 6 above)
Comment (anukul) [reply storyruntime/processing/Lexicon.py]: You mean this?
```py
if isinstance(data, list):
    value = b
else: # isinstance(data, dict), but no elif for full coverage
    value = a
story.set_variable(assign={'paths': output}, output=value)
```

Comment (anukul) [reply storyruntime/processing/Lexicon.py]: If we write the `for` block twice, we end up duplicating everything else inside :/

Comment (wilzbach) [reply storyruntime/processing/Lexicon.py]: 1) Nah you don't need to have everything inside the loop like e.g. this doesn't need to be run on every iteration

```py 
          output = line['output']
          assert 1 <= len(output) <= 2, \
              f'foreach output must be 1 or 2 values, found {len(output)}'
```

2) I don't see how this is a huge duplication

```py
          if isinstance(data, list):
                foreach i, v in enumerate(data):
                     if len(output) == 1:
                         set_output(v)
                    else:
                         set_output(i, v)
```

Also, it's so much more explicit than `a` + `b`.
And if you really want to save line number, you can use arrays and do e.g. `set_output([i] if len(output) == 1 else [i,v])`.

>  we end up duplicating everything else inside

This is good here, because with the duplication you would have noticed that maps with only one output aren't covered.

Comment (anukul) [reply storyruntime/processing/Lexicon.py]: > I don't see how this is a huge duplication
>```py
>          if isinstance(data, list):
>                foreach i, v in enumerate(data):
>                     if len(output) == 1:
>                         set_output(v)
>                    else:
>                         set_output(i, v)
>```
The duplication also includes what comes after `set_output`... `execute_block`, `break`, `continue`

> This is good here, because with the duplication you would have noticed that maps with only one output aren't covered.

Sorry I don't understand, what's not covered at the moment?

File: tests/integration/processing/Lexicon.py [hunk 7]
```diff
@@ -471,6 +471,54 @@
             )
         ]
     ),
+    Suite(
+        preparation_lines='l = [8, 9, 10]\n'
+                          'index_sum = 0\n'
+                          'value_sum = 0\n'
+                          'foreach l as index, value\n'
```
Comment (judepereira) [inline tests/integration/processing/Lexicon.py]: @anukul If I just specify `foreach l as item`, is item the index or an actual element in the list `l`?

Comment (anukul) [reply tests/integration/processing/Lexicon.py]: It is the list element
